src/include/ipc/probe.h

```diff
@@ -0,0 +1,182 @@
+#define PROBE_MASK_SAMPLE_FMT		MASK(9, 9)
+#define PROBE_MASK_SAMPLE_END		MASK(8, 8)
+#define PROBE_MASK_INTERLEAVING_ST	MASK(7, 7)
+
+/**
+ * Header for data packets sent via compressed PCM from extraction probes
+ */
+struct probe_data_packet {
+	uint32_t sync_word;		/**< PROBE_EXTRACT_SYNC_WORD */
+	uint32_t buffer_id;		/**< Buffer ID from which data was extracted */
+	uint32_t format;		/**< Encoded data format */
+	uint32_t timestamp_low;		/**< Low 32 bits of timestamp in us */
+	uint32_t timestamp_high;	/**< High 32 bits of timestamp in us */
+	uint32_t checksum;		/**< CRC32 of header and payload */
+	uint32_t data_size_bytes;	/**< Size of following audio data */
+	uint32_t data[];		/**< Audio data extracted from buffer */
+} __attribute__((packed));
+
+/**
+ * Description of probe dma
+ */
+struct probe_dma {
+	uint32_t stream_tag;		/**< Stream tag associated with this DMA */
+	uint32_t dma_buffer_size;	/**< Size of buffer associated with this DMA */
+} __attribute__((packed));
+
+/**
+ * Description of probe point
+ */
+struct probe_point {
+	uint32_t buffer_id;	/**< ID of buffer to which probe is attached */
+	uint32_t purpose;	/**< PROBE_PURPOSE_EXTRACTION or PROBE_PURPOSE_INJECTION */
+	uint32_t stream_tag;	/**< Stream tag of DMA via which data will be provided for injection.
+				 *   For extraction purposes, stream tag is ignored when received,
+				 *   but returned actual extraction stream tag via INFO function.
+				 */
+} __attribute__((packed));
+
+/**
+ * \brief DMA ADD for probes.
+ *
+ * Size of array is inferred from the reported IPC header size.
+ *
+ * Used as payload for IPCs: SOF_IPC_PROBE_INIT, SOF_IPC_PROBE_DMA_ADD.
+ */
+struct sof_ipc_probe_dma_add_params {
+	struct sof_ipc_cmd_hdr hdr;	/**< Header */
+	struct probe_dma probe_dma[];	/**< Array of DMAs to be added */
+} __attribute__((packed));
+
+/**
+ * \brief Reply to INFO functions.
+ *
+ * Size of array is inferred from the reported IPC header size.
+ *
+ * Used as payload for IPCs: SOF_IPC_PROBE_DMA_INFO, SOF_IPC_PROBE_POINT_INFO.
+ */
+struct sof_ipc_probe_info_params {
+	struct sof_ipc_reply rhdr;			/**< Header */
+	union {
+		struct probe_dma probe_dma[0];		/**< DMA info */
+		struct probe_point probe_point[0];	/**< Probe Point info */
+	};
+} __attribute__((packed));
+
+/**
+ * \brief Probe DMA remove.
+ *
+ * Size of array is inferred from the reported IPC header size.
+ *
+ * Used as payload for IPC: SOF_IPC_PROBE_DMA_REMOVE
+ */
+struct sof_ipc_probe_dma_remove_params {
+	struct sof_ipc_cmd_hdr hdr;	/**< Header */
+	uint32_t stream_tag[];		/**< Array of stream tags associated with DMAs to remove */
+} __attribute__((packed));
+
+/**
+ * \brief Add probe points.
+ *
+ * Size of array is inferred from the reported IPC header size.
+ *
+ * Used as payload for IPC: SOF_IPC_PROBE_POINT_ADD
+ */
+struct sof_ipc_probe_point_add_params {
+	struct sof_ipc_cmd_hdr hdr;		/**< Header */
+	struct probe_point probe_point[];	/**< Array of Probe Points to add */
```

Comment:
Another point I just thought of: for IPC structures, we try to add reserved fields so that we provide backwards-compatible solutions if in the future we need to describe additional features.
With the proposal where the header size is used to find out how many probe points or stream tags are used, is there even a line of sight for such minor ABI changes?

Reply:
Well, nope. But honestly, this interface has already passed the trial of time in cAVS project. Also, this is a development/debug feature, so I believe that in such context, backward compatibility has slightly less of a weight attached to it.

Reply:
I wonder how many additional holes I am going to find at this rate...

@lgirdwood what do you think on this one?

Reply:
ack, lets not use hdr.size to be the number of array elems, this should be separate member.

Reply:
@lgirdwood but that's contrasting with initial requirement to take cAVS interface as-is, without modification, other than project-specific IPC headers. Is that point still valid?

Reply:
the intent was to use things without modification, but in this example there are issues/limitations, so the 'as is' is no longer applicable in my opinion. converging solutions is a two-way street.

Reply:
@lgirdwood btw. as I've written before, we're not putting array members count into `hdr.size`, because that would be a painfully stupid thing to do. We're inferring that, based on real size of the payload which is properly placed in `hdr.size`.
As @plbossart stated, that's limping on backward compatibility and protection against invalid IPC packets, but it's the same thing that's done in cAVS and seems to fare just ok.